ffi_internals/src/impl_internals/fn_ffi.rs

```diff
@@ -21,39 +42,129 @@ pub struct FnFFI {
+            self.local_aliases
+                .get(&type_path.path.segments.last().unwrap().ident)
+                .map_or_else(|| ty.deref().clone(), std::clone::Clone::clone)
+        } else {
+            ty.deref().clone()
+        }
+    }
+}
+
+impl<'a> From<FnFFIInputs<'a>> for FnFFI {
+    fn from(inputs: FnFFIInputs<'_>) -> Self {
+        let fn_name = inputs.method.sig.ident.clone();
+        let (arguments, receiver) = inputs.method.sig.inputs.iter().fold(
+            (Vec::<FnParameterFFI>::new(), FnReceiver::None),
+            |mut acc, input| {
+                match input {
+                    syn::FnArg::Receiver(receiver) => {
+                        acc.1 = if receiver.reference.is_some() {
+                            FnReceiver::Borrowed
+                        } else {
+                            FnReceiver::Owned
+                        }
+                    }
+                    syn::FnArg::Typed(arg) => acc.0.push(FnParameterFFI::from((
+                        arg,
+                        inputs.raw_types.clone(),
+                        Some(inputs.self_type.clone()),
+                    ))),
+                }
+                acc
+            },
+        );
+
+        let return_type: Option<TypeFFI> = match &inputs.method.sig.output {
+            syn::ReturnType::Default => None,
+            syn::ReturnType::Type(_token, ty) => {
+                let dealiased = inputs.strip_local_alias(&*ty);
+                Some(TypeFFI::from(TypeAttributes::initial(
+                    dealiased,
+                    inputs.raw_types,
+                    Some(inputs.self_type),
+                )))
+            }
+        };
+
+        Self {
+            fn_name,
+            receiver,
+            parameters: arguments,
+            return_type,
+        }
+    }
 }
 
-impl From<&ImplItemMethod> for FnFFI {
-    fn from(method: &ImplItemMethod) -> Self {
+impl From<(&ItemFn, Vec<Ident>)> for FnFFI {
+    fn from(data: (&ItemFn, Vec<Ident>)) -> Self {
+        let (method, raw_types) = data;
         let fn_name = method.sig.ident.clone();
-        let (arguments, has_receiver) = method.sig.inputs.iter().fold(
-            (Vec::<FnParameterFFI>::new(), false),
+        let (arguments, receiver) = method.sig.inputs.iter().fold(
+            (Vec::<FnParameterFFI>::new(), FnReceiver::None),
             |mut acc, input| {
                 match input {
-                    syn::FnArg::Receiver(_receiver) => acc.1 = true,
-                    syn::FnArg::Typed(arg) => acc.0.push(FnParameterFFI::from(arg)),
+                    syn::FnArg::Receiver(receiver) => {
+                        acc.1 = if receiver.reference.is_some() {
+                            FnReceiver::Borrowed
+                        } else {
+                            FnReceiver::Owned
+                        }
+                    }
+                    // TODO: We should use inputs.strip_local_alias(type) on parameters, too, since
```

Comment:
Do this (and see if we can de-dupe this with the `From<FnFFIInputs>` impl above).

Reply:
NM, this is a different context. Local aliases are only relevant in the context of an impl; we use `FnFFIInputs` in that context to capture all of that additional data. This is for the simpler case of exposing an unassociated function from a module -- it's possible to have local aliases to worry about in this case, but that would need a bunch of separate work (and probably a helper attribute since they're defined outside of the tokens we're operating on) to support.